lib/pbench/cli/agent/commands/tools/clear.py

```diff
@@ -25,83 +25,105 @@ def __init__(self, context):
-            return 1
-
-        try:
-            remotes = self.context.remote.split(",")
-        except Exception:
-            # We were not given any remotes on the command line, build the list
-            # from the tools group directory.
-            remotes = self.remote(self.tool_group_dir)
-            if not remotes:
-                # FIXME:  This is a weird case -- the group directory exists,
-                #  but contains no remotes. This is the result of clearing the
-                #  tools from a group which has already been cleared, because
-                #  the first clear does not remove the group directory.  This
-                #  behavior might be desirable for the `default` directory, but
-                #  it seems wrong for custom groups.
-                self.logger.error(f'No such group "{self.context.group}".')
-                return 1
-
-        tool_not_found = True  # Not found, yet
-        for remote in remotes:
-            tg_dir_r = self.tool_group_dir / remote
-            if not tg_dir_r.exists():
-                self.logger.warn(
-                    'No remote host "%s" in group %s.', remote, self.context.group,
-                )
+        groups = self.context.group.split(",")
+        for group in groups:
+            if self.verify_tool_group(group) != 0:
+                self.logger.error(f'No such group "{group}".')
+                errors = 1
                 continue
 
-            if self.context.name:
-                names = [self.context.name]
-            else:
-                # Discover all the tools registered for this remote
-                names = self.tools(tg_dir_r)
-                if not names:
-                    # FIXME:  this is another odd case -- the remote subdirectory
-                    #  exists, but it's empty.  (We'll remove it below.)
+            try:
+                remotes = self.context.remote.split(",")
+            except Exception:
+                # We were not given any remotes on the command line, build the list
+                # from the tools group directory.
+                remotes = self.remote(self.tool_group_dir)
+                if not remotes:
+                    # FIXME:  This is a weird case -- the group directory exists,
+                    #  but contains no remotes. This is the result of clearing the
+                    #  tools from a group which has already been cleared, because
+                    #  the first clear does not remove the group directory.  This
+                    #  behavior might be desirable for the `default` directory, but
+                    #  it seems wrong for custom groups.
+                    try:
+                        if group != "default":
+                            shutil.rmtree(self.tool_group_dir)
+                    except OSError:
+                        self.logger.error(
+                            "Failed to remove group directory %s", self.tool_group_dir
```

Comment:
Maybe "Failed to remove empty group directory %s"??

As I mentioned in Webb's review, I think talking about "group directories" exposes unnecessary internals, but then the whole concept of this error is inherently internal and I suspect we should just clean it up and ignore it silently... or even complain that they're trying to clear a group that doesn't exist.

Reply:
I'm opposed to failing silently.  And, in this case, I think we need to report enough information to the user that s/he can proceed to diagnose and possibly resolve the problem.

I aggree that the fact that "groups" are implemented as directories should largely be hidden from the user.  But, the root of this failure is that we need to remove a directory because we ostensibly created it (we shouldn't be leaving junk behind), and we've found that we cannot do that (which is a surprise).

So, I think we need to report that we wanted to remove the directory, what the directory's name is, and, optimally, why the removal failed (i.e., we should append the stringized exception object).  This should allow the user to remove it him-/herself (or to request that a sysadmin remove it).

I have no objection to pointing out that it is a "group directory", but I don't think we should ignore it silently.  If the group doesn't exist, I think we should detect and handle that in a different way (but, yeah, we should complain about that, too).